src/gil.rs

```diff
@@ -467,7 +478,10 @@ pub unsafe fn register_decref(obj: NonNull<ffi::PyObject>) {
     if gil_is_acquired() {
         ffi::Py_DECREF(obj.as_ptr())
     } else {
+        #[cfg(feature = "reference-pool")]
         POOL.register_decref(obj);
+        #[cfg(all(not(feature = "reference-pool"), debug_assertions))]
+        eprintln!("Leaking pointer into Python heap because the GIL is not held.");
```

Comment:
Is it potentially helpful to give the address of the pointer?

Reply:
I think if finding such an issue, then a backtrace would be relevant but if we go for aborting that would result in any case.

> But I would be open to aborting instead of printing :)

I am fine with making that change. We just need to document this with a big red warning sign but as you wrote, this should get its own section anyway.

As for printing the pointer value, I think this can become a security issue as pointer values can be used to break ASLR and with services exposed to the web, it is conceivable that such a message ends up with the attacker. This would be a security issue for itself but following a defence in depth approach, we might want to be careful about printing pointer values.

Reply:
One last problem I have is that `std::process::abort` does not take a message and there is no stable portable way to do it ourselves and we do not have the GIL to e.g. call the unraisable hook.

For `PanicTrap` we just went for explicitly causing a double panic by panicking in `Drop` which is only called during unwinding. This was also was gave us the backtrace.

I shoehorned the same approach here but left it as a separate commit so we can decide to drop it if this deemed too hacky.